soroban-env-host/src/host.rs

```diff
@@ -1114,24 +1132,53 @@ impl VmCallerCheckedEnv for Host {
     }
 
     // Notes on metering: covered by the components
+    /// TODO: Rename to get_source. https://github.com/stellar/rs-soroban-env/issues/470
     fn get_invoking_contract(&self, _vmcaller: &mut VmCaller<Host>) -> Result<Object, HostError> {
         let frames = self.0.context.borrow();
         // the previous frame must exist and must be a contract
-        let hash: Hash = if frames.len() >= 2 {
+        let contract_id: Option<Hash> = if frames.len() >= 2 {
             match &frames[frames.len() - 2] {
                 #[cfg(feature = "vm")]
-                Frame::ContractVM(vm, _) => Ok(vm.contract_id.metered_clone(&self.0.budget)?),
-                Frame::HostFunction(_) => {
-                    Err(self.err_general("Host function context has no contract ID"))
-                }
-                Frame::Token(id, _) => Ok(id.clone()),
+                Frame::ContractVM(vm, _) => Some(vm.contract_id.metered_clone(&self.0.budget)?),
+                Frame::HostFunction(_) => None,
+                Frame::Token(id, _) => Some(id.clone()),
+                #[cfg(feature = "testutils")]
+                Frame::TestContract(id, _) => Some(id.clone()), // no metering
+            }
+        } else {
+            None
+        };
+        let id = if let Some(contract_id) = contract_id {
+            <Vec<u8>>::from(contract_id.0)
+        } else if let Some(Source::Account(AccountId(PublicKey::PublicKeyTypeEd25519(Uint256(
+            account_id,
+        ))))) = &*self.0.source.borrow()
+        {
+            <Vec<u8>>::from(*account_id)
+        } else {
+            return Err(self.err_general("no invoking account or contract"));
+        };
+        Ok(self.add_host_object::<Vec<u8>>(id)?.into())
+    }
+
+    // TODO: Assess metering.
+    fn get_source_type(&self, _vmcaller: &mut VmCaller<Host>) -> Result<u32, HostError> {
```

Comment:
I don't love duplicating this rather complex matching process, but to move both implementations into a single location will cause a double allocation. Open to feedback on how we can improve this.